pkg/ephemeral/io/tuple_streamer.go

```diff
@@ -155,39 +153,65 @@ func NewCastorTupleStreamerWithWriterFactory(l *zap.SugaredLogger, tt castor.Tup
 
 // CastorTupleStreamer provides tuples to the SPDZ execution for the given type and configuration.
 type CastorTupleStreamer struct {
-	logger        *zap.SugaredLogger
-	pipeWriter    PipeWriter
-	tupleType     castor.TupleType
-	stockSize     int32
-	castorClient  castor.AbstractClient
-	baseRequestID uuid.UUID
-	requestCycle  int
-	headerData    []byte
-	streamData    []byte
+	logger         *zap.SugaredLogger
+	pipeWriter     PipeWriter
+	tupleType      castor.TupleType
+	stockSize      int32
+	castorClient   castor.AbstractClient
+	baseRequestID  uuid.UUID
+	requestCycle   int
+	headerData     []byte
+	streamData     []byte
+	streamerDoneCh chan struct{}
+	tupleBufferCh  chan []byte
+	fetchTuplesCh  chan struct{}
+	// bufferLckCh is used as a synchronization lock, where on routine can lock the channel by writing to it. Each
+	// consecutive write will block the writing routine until the channel has been unlocked by reading from it. In
+	// combination with es "select" statement, the channel can be used as a "timeout-able" lock.
+	//
+	// Reading is supposed to be performed by the initial routine which wrote to the channel.
+	bufferLckCh   chan struct{}
 	streamedBytes int
 }
 
 // StartStreamTuples repeatedly downloads a given type of tuples from castor and streams it to the according file as
 // required by MP-SPDZ
 func (ts *CastorTupleStreamer) StartStreamTuples(terminateCh chan struct{}, errCh chan error, wg *sync.WaitGroup) {
 	ts.streamData = append(ts.streamData, ts.headerData...)
-	pipeWriterReady := make(chan struct{})
+	ts.streamerDoneCh = make(chan struct{})
+	ts.fetchTuplesCh = make(chan struct{}, 1)
+	ts.bufferLckCh = make(chan struct{}, 1)
+	ts.tupleBufferCh = make(chan []byte, 1)
+	ts.fetchTuplesCh <- struct{}{}
 	go func() {
 		defer func() {
-			var streamedTupleBytes, discardedTupleBytes int
+			close(ts.streamerDoneCh)
+			select {
+			case ts.bufferLckCh <- struct{}{}:
+			// wait for buffer routine and block
```

Comment:
Maybe you can add to that comment, why we are waiting here.

Reply:
done